### ISSUE TITLE
`<variant>`: Should we treat P0608R3 as a DR against C++17?

### ISSUE THREAD

**Issue by frederick-vs-ja** [microsoft/STL#4412]
WG21-P0608R3 was intentedly not treated as DR by MSVC STL (https://github.com/microsoft/STL/pull/1629#issuecomment-778895630).

However, the author of P0608R3 (@zhihaoy) applied those changes to libc++ in C++17 mode (LLVM-D44865), and so did libstdc++ (https://github.com/gcc-mirror/gcc/commit/d069df01eda5094cb2c934a7e1567219ba36e511). So it's arguably that P0608R3 should be treated as DR even though it's not so treated officially.

There's at least one user even thinking MSVC STL is buggy (see https://gcc.gnu.org/bugzilla/show_bug.cgi?id=113007#c7).

**Comment by jwakely** [comment]
See also https://github.com/microsoft/STL/issues/2171#issuecomment-908715926

**Comment by toughengineer** [comment]
> There's at least one user even thinking MSVC STL is buggy (see https://gcc.gnu.org/bugzilla/show_bug.cgi?id=113007#c7).

That user is me.
Just to clarify, I wrote in that bug conversation:
> Now it's time to file a bug in MSVC's standard library.

but before actually filing a bug I learned roughly from the issue mentioned by @jwakely above that MSVC developers decided to leave the initial implementation in C++17 mode for backwards compatibility, at least this is my understanding, so I never filed an issue because I though it is redundant.

**Comment by StephanTLavavej** [comment]
We talked about this at the weekly maintainer meeting, and we just don't have the maintainer capacity needed to deal with cleaning up usage in the compiler and potentially Windows and elsewhere, if we changed our decision here to align with libstdc++ and libc++. Our behavior is conforming according to C++17 and we need to spend our energy elsewhere.

**Comment by StephanTLavavej** [comment]
Revisiting this - it turned out to be necessary to fully update our libcxx-derived `variant` tests. I'll have a PR in the near-ish future.

**Comment by toughengineer** [comment]
If you decide to change the behavior in C++17 mode to include P0608, please consider updating macros
https://github.com/microsoft/STL/blob/e36ee6c2b9bc6f5b1f70776c18cf5d3a93a69798/stl/inc/yvals_core.h#L881-L883
`_CPPLIB_VER` (currently 650, as in ?)
`_MSVC_STL_VERSION` (currently 143, as in runtime version 14.3/v143?)
`_MSVC_STL_UPDATE` (I guess only this one will be updated)
so that user code can detect that starting with _that_ version of the standard library it does not need a workaround for unimplemented P0608.

**Comment by frederick-vs-ja** [comment]
> `_MSVC_STL_UPDATE` (I guess only this one will be updated)

Yeah, this one is updated every month. We regularly opens a good-first-issue for updating it.
